File: Sources/Services/Security/AgentEnvironmentBuilder.swift
// MARK: - AgentEnvironmentBuilder
// Builds an allowlist-based environment for AI CLI agent subprocesses.
// macOS 14+, Swift 5.10

import Foundation

/// Constructs a minimal, safe environment for launching AI CLI agents
/// in a dedicated PTY.
///
/// Only variables from an explicit allowlist are inherited from the parent
/// process. Sensitive variables (`AWS_*`, `DATABASE_URL`, `*_SECRET`,
/// `*_TOKEN`, `*_PASSWORD`, `GITHUB_*`) are never forwarded.
///
/// The agent's API key is injected explicitly from Keychain or environment.
enum AgentEnvironmentBuilder {

    /// Variables safe to forward to agent subprocesses.
    private static let allowedVariables: Set<String> = [
        "HOME", "USER", "LOGNAME",
        "LANG", "LC_ALL", "LC_CTYPE",
        "TERM", "COLORTERM",
        "PATH", "SSH_AUTH_SOCK",
        "SHELL", "TMPDIR",
        "XDG_CONFIG_HOME", "XDG_DATA_HOME",
        "OPENAI_API_KEY", "ANTHROPIC_API_KEY",
    ]

    /// Build an environment array (`["KEY=VALUE", ...]`) for the given agent.
    ///
    /// - Parameters:
    ///   - assistant: The AI assistant to launch.
    ///   - apiKeyValue: The API key value to inject (from Keychain or env).
    ///     Pass `nil` if the agent does not require an API key.
    /// - Returns: Array of `"KEY=VALUE"` strings suitable for `startProcess(environment:)`.
    static func build(for assistant: AIAssistant, apiKeyValue: String?) -> [String] {
        let processEnv = ProcessInfo.processInfo.environment
        var result: [String: String] = [:]

        for key in allowedVariables {
            if let value = processEnv[key] {
                result[key] = value
            }
        }

        // Ensure terminal capabilities are always set for proper rendering.
        result["TERM"] = "xterm-256color"
        result["COLORTERM"] = "truecolor"
        result["LANG"] = result["LANG"] ?? "en_US.UTF-8"

        // Agent subprocesses run without a login shell and don't benefit from
        // .zprofile PATH augmentation. When VibeStudio is launched from Finder/Dock,
        // launchd provides a minimal PATH (/usr/bin:/bin:/usr/sbin:/sbin) that
        // excludes Homebrew, npm-global, cargo, etc.
        //
        // We prepend the same trusted directories used by CLIAgentPathResolver so
        // the agent binary can locate itself (for self-invocation, updates) and
        // other tools (git, node) without relying on the parent process PATH.
        // Sourced from SecurityConstants.trustedBinDirectories — single source of truth.
        let trustedBins = SecurityConstants.trustedBinDirectories
        let currentPath = result["PATH"] ?? "/usr/bin:/bin:/usr/sbin:/sbin"
        let existingParts = currentPath.split(separator: ":").map(String.init)
        let missingBins = trustedBins.filter { !existingParts.contains($0) }
        if !missingBins.isEmpty {
            result["PATH"] = (missingBins + existingParts).joined(separator: ":")
        }

        // Inject the agent-specific API key if provided.
        if let envVar = assistant.apiKeyEnvironmentVariable,
           let key = apiKeyValue,
           !key.isEmpty {
            result[envVar] = key
        }

        return result.map { "\($0.key)=\($0.value)" }
    }
}

File: shared/src/commonMain/kotlin/studio/vibe/shared/service/security/AgentEnvironmentBuilder.kt
package studio.vibe.shared.service.security

import studio.vibe.shared.constants.SecurityConstants
import studio.vibe.shared.contract.AIAgent

/**
 * Constructs a minimal, safe environment map for launching AI CLI agents.
 *
 * Only variables from an explicit allowlist are inherited from the parent process
 * environment. Sensitive variables (AWS_*, DATABASE_URL, *_SECRET, *_TOKEN,
 * *_PASSWORD, GITHUB_*) are never forwarded because they are not in the allowlist.
 *
 * Returns Map<String, String>. Callers that need "KEY=VALUE" strings for PTY/process
 * launch must format the map themselves (platform-specific concern).
 */
object AgentEnvironmentBuilder {

    /** Variables safe to forward to agent subprocesses. */
    private val allowedVariables: Set<String> = setOf(
        "HOME", "USER", "LOGNAME",
        "LANG", "LC_ALL", "LC_CTYPE",
        "TERM", "COLORTERM",
        "PATH", "SSH_AUTH_SOCK",
        "SHELL", "TMPDIR",
        "XDG_CONFIG_HOME", "XDG_DATA_HOME",
        "OPENAI_API_KEY", "ANTHROPIC_API_KEY",
    )

    /**
     * Build an environment map for the given agent.
     *
     * @param assistant   The AI assistant to launch.
     * @param apiKeyValue The API key value to inject (from Keychain or env).
     *   Pass null if the agent does not require an API key.
     * @param currentEnv  The current process environment (pass ProcessEnvironment on each platform).
     * @return Map of environment variable names to values.
     */
    fun build(
        agent: AIAgent,
        apiKeyValue: String?,
        currentEnv: Map<String, String>,
    ): Map<String, String> {
        val result = mutableMapOf<String, String>()

        // Copy allowlisted variables from the current environment.
        for (key in allowedVariables) {
            val value = currentEnv[key]
            if (value != null) {
                result[key] = value
            }
        }

        // Ensure terminal capabilities are always set for proper rendering.
        result["TERM"] = "xterm-256color"
        result["COLORTERM"] = "truecolor"
        if (!result.containsKey("LANG")) {
            result["LANG"] = "en_US.UTF-8"
        }

        // Agent subprocesses run without a login shell and don't benefit from
        // .zprofile PATH augmentation. Prepend trusted directories so the agent
        // binary can locate itself and other tools (git, node) without relying
        // on the parent process PATH. Sourced from SecurityConstants — single
        // source of truth.
        val homeDir = result["HOME"] ?: ""
        val trustedBins = SecurityConstants.trustedBinDirectories(homeDir)
        val currentPath = result["PATH"] ?: "/usr/bin:/bin:/usr/sbin:/sbin"
        val existingParts = currentPath.split(":").filter { it.isNotEmpty() }
        val missingBins = trustedBins.filter { it !in existingParts }
        if (missingBins.isNotEmpty()) {
            result["PATH"] = (missingBins + existingParts).joinToString(":")
        }

        // Inject the agent-specific API key if provided.
        val envVar = agent.apiKeyEnvironmentVariable
        if (envVar != null && !apiKeyValue.isNullOrEmpty()) {
            result[envVar] = apiKeyValue
        }

        return result
    }
}
